pkg/tagger/tagstore.go

```diff
@@ -131,31 +137,64 @@ func (e *entityTags) get(highCard bool) ([]string, []string) {
 	}
 
 	// Cache miss
-	var arrays [][]string
 	var sources []string
-	lowCardCount := 0
+	tagSortList := make(map[string][]tagSort)
 
 	for source, tags := range e.lowCardTags {
-		arrays = append(arrays, tags)
-		lowCardCount += len(tags)
 		sources = append(sources, source)
+		insertWithPriority(tagSortList, tags, source, false)
 	}
-	for _, tags := range e.highCardTags {
-		arrays = append(arrays, tags)
+
+	for source, tags := range e.highCardTags {
+		insertWithPriority(tagSortList, tags, source, true)
+	}
+
+	lowCardTags := []string{}
+	highCardTags := []string{}
+	for _, tags := range tagSortList {
+		for i := 0; i < len(tags); i++ {
+			insert := true
+			for j := 0; j < len(tags); j++ {
+				// if we find a duplicate tag with higher priority we do not insert the tag
+				if i != j && tags[i].priority < tags[j].priority {
+					insert = false
+					break
+				}
+			}
+			if insert {
+				if tags[i].isHighCard {
+					highCardTags = append(highCardTags, tags[i].tag)
+				} else {
+					lowCardTags = append(lowCardTags, tags[i].tag)
+				}
+			}
+		}
 	}
-	tags := utils.ConcatenateTags(arrays)
+
+	tags := append(lowCardTags, highCardTags...)
 
 	// Write cache
 	e.RUnlock()
 	e.Lock()
 	e.cacheValid = true
 	e.cachedSource = sources
 	e.cachedAll = tags
-	e.cachedLow = e.cachedAll[:lowCardCount]
+	e.cachedLow = e.cachedAll[:len(lowCardTags)]
 	e.Unlock()
 
 	if highCard {
 		return tags, sources
 	}
-	return tags[:lowCardCount], sources
+	return lowCardTags, sources
+}
+
+func insertWithPriority(tagSortList map[string][]tagSort, tags []string, source string, isHighCard bool) {
+	for _, t := range tags {
+		tagName := strings.Split(t, ":")[0]
+		tagSortList[tagName] = append(tagSortList[tagName], tagSort{
+			tag:        t,
+			priority:   collectors.CollectorPriorities[source],
```

Comment:
The priority lookup could be moved before the loop, with a big fat warnf if the lookup fails